functiontrace: add functiontrace-server dependency

tehmatt opened the pull request:
###### Description of changes

This is needed for `functiontrace foo.py` to work, as `functiontrace-server` must be in your path.

###### Things done


- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2082

ocfox commented:
duplicate #222970

tehmatt commented:
Hmm, I'll resolve that.  I suspect this approach should take precedence.  
How do we generally feel about breaking changes (moving packages)?

ocfox commented:
What's mean? If you want to fix or change this package, you can do it on file which has been merged. No need to `moving packages`.

tehmatt commented:
Functiontrace isn't really a python package, but is instead a version-agnostic tool that works with Python programs, so a top-level `functiontrace` makes more sense than `python3XPackages.functiontrace`.

tehmatt committed:
functiontrace: add functiontrace-server dependency

This is needed for `functiontrace foo.py` to work, as
`functiontrace-server` must be in your path.

tehmatt commented:
I think there's a decent argument that this should still be a `pythonPackage`, so I updated this PR to work with that module.

tehmatt commented on pkgs/development/python-modules/functiontrace/default.nix:
```diff
@@ -15,17 +10,21 @@ buildPythonPackage rec {
   pythonImportsCheck = [ "functiontrace" ];
 
+  # `functiontrace` needs `functiontrace-server` in its PATH.
+  propagatedBuildInputs = [ functiontrace-server ];
```
I'm not sure how to test that this works on import.  The `postInstall` wrapper certainly does, and that's the much more common way to run `functiontrace` at least.

ocfox replied:
Even with this method, I don't think `propagatedBuildInputs` should be used here.

SuperSandro2000 replied:
we either need propagatedBuildInputs or makeWrapper, not both. Also please revert the formatting

tehmatt replied:
Updated to use only `propagatedBuildInputs`.

ocfox replied:
```suggestion
  postInstall = ''
    wrapProgram $out/bin/functiontrace \
      --set PATH ${lib.makeBinPath [ functiontrace-server ]}
  '';
```
I think this is the more logical way, `functiontrace-server` is a rust package and should not be added to python propagatedBuildInputs.

SuperSandro2000 replied:
Yes, like ocfox said.

ocfox commented:
Result of `nixpkgs-review pr 224570` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.functiontrace</li>
    <li>python310Packages.functiontrace.dist</li>
    <li>python311Packages.functiontrace</li>
    <li>python311Packages.functiontrace.dist</li>
  </ul>
</details>
